Add `is.numericString()`

itaisteinherz opened the pull request:
I implemented the method based on the discussion over at #54 (and this closes #54).
Note: I added the check `!nullOrUndefined(value)`, since without it `is.numericString(null)` would return `true` (because `Number(null)` is `0`).

This is the first time I ever used TypeScript, and so I'd love to get some feedback on this PR.

gioragutt commented on source/index.ts:
```diff
@@ -135,10 +135,13 @@ namespace is { // tslint:disable-line:no-namespace
 	// tslint:enable:variable-name
 
+	export const nullOrUndefined = (value: any): value is null | undefined => null_(value) || undefined(value);
+	export const numericString = (value: any): value is String =>
```
if this is supposed to be a `numericString` how come you accept `any`? and how does it indicate that `value is String`?
First, use `string` over `String`, second - I believe this is not the correct signature for this method.
I think that it should just return bool, since you can't really create a type guard that assures you that `string` is typed as a `number`, you can just have a bool that tells you that, and you can just `Number(_)` it...
```suggestion
	export const numericString = (value: string | number): bool =>
```

the `|  number` is optional, only if you feel that's necessary...

itaisteinherz commented:
@gioragutt Thanks for the feedback! I've fixed the method's signature as per your suggestions.
The method currently only accepts a `string`, since accepting a `number` as well seems pointless to me.

Also, as some comments in #54 indirectly discuss the result of running `is.numericString(undefined)`, maybe we should revert the method's parameter to be `any`.

@sindresorhus What do you think?

gioragutt commented on source/index.ts:
```diff
@@ -136,8 +136,8 @@ namespace is { // tslint:disable-line:no-namespace
-	export const numericString = (value: any): value is String =>
-		!Number.isNaN(Number(value)) && !nullOrUndefined(value);
+	export const numericString = (value: string): bool =>
+		!Number.isNaN(Number(value));
```
why did you remove the `nullOrUndefined` check?

itaisteinherz replied:
I removed it seeing that if `value` is either `null` or `undefined`, the TS compiler would immediately return an error (since the parameter is not a `string`).
In case we revert to accepting `any`, the `nullOrUndefined` check would be required again.

gioragutt replied:
Got it: add a `is.string` check. and then you can change `value` back to being `any`.

```ts
export const numericString = (value: any) =>
    is.string(value) && !Number.isNaN(Number(value));
```

brandon93s replied:
I believe the intent here is to check if a variable contains any representation of a number:

```js
if (is.number(val) || is.numericString(val)) {
   console.log(`The unicorn is ${val}`)
}
```

With that I mind, I it should accept `any` and type check `string` internally.
```

sindresorhus replied:
Yes, it should accept `any`, like all the other checks.

brandon93s commented on source/tests/test.ts:
```diff
@@ -426,6 +426,12 @@ test('is.symbol', t => {
+test('is.numericString', t => {
+	t.true(m.numericString('5'));
+	t.true(m.numericString('-3.2'));
+	t.true(m.numericString('Infinity'));
```
The following test case should not pass, but does:

```js
t.true(m.numericString(''));
```

```js
Number('')
// => 0
```

We need to guard against empty string.

gioragutt replied:
Why not use `testType` here as well, instead of just checking several specific use cases?

The tests wouldn't even break here if you replaces the implementation to `return true;`.

itaisteinherz commented:
@brandon93s I've changed the type of `value` to `any`, fixed the issue you pointed out and added a test for it.
@gioragutt I've moved to using `testType` (I tried to use it before I even opened the PR but couldn't because there was a problem I haven't figured out until now).

gioragutt commented:
@itaisteinherz do you need help with working something out? We can talk over WhatsApp if you need (Israeli here as well)

sindresorhus commented on readme.md:
```diff
@@ -96,6 +96,7 @@ All the below methods accept a value and returns a boolean for whether the value
 
 Keep in mind that [functions are objects too](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions).
 
+##### .numericString(value)
```
There should be a description here, as it's not immediately clear what a `numericString` entails.

sindresorhus commented on readme.md:
```diff
@@ -96,6 +96,10 @@ All the below methods accept a value and returns a boolean for whether the value
 
+##### .numericString(value)
+
+Returns `true` for a string that contains a number.
```
That's not entirely correct. It returns true for a string that **represents** a number. If it was *contains*, then this would be valid `is.numericString('foo 4 bar') //=> true`.

joker314 commented:
This was a really good PR!

Could I just note that it's not entirely clear that the string `"NaN"` should be rejected (given that `NaN` is of type number, and `"Infinity"` and "-Infinity"` aren't. Should this be documented?

itaisteinherz commented:
> Could I just note that it's not entirely clear that the string `"NaN"` should be rejected (given that `NaN` is of type number, and `"Infinity"` and `"-Infinity"` aren't. Should this be documented?

I personally don't think so, since we specifically check in the source code if the given `value` is not `NaN` (see [here](https://github.com/sindresorhus/is/pull/67/files#diff-21e9ddd93162651bd36f6e5bbfca8460R140)).

joker314 commented:
People don't read the source code of the libraries they use, normally.

Instead, they refer to documentation (that's the README.md file here). The documentation just says that it returns true for a "string representing a number".

There are different meanings of the word "number". In JavaScript, `NaN`, `10`, `Infinity` are all numbers. The function introduced in these changes considers `"Infinity"` and `"10"` numbers, but it doesn't consider `"NaN"` a number. I'm not criticising this design choice 🙂, instead, I'm asking if it's a good idea for this design choice to be noted down for developers to see.

Not putting something in the documentation just because the code is self-explanatory isn't a good idea.

```javascript
typeof NaN === 'number' // true
```

itaisteinherz commented:
Opinions? @sindresorhus @gioragutt @brandon93s

sindresorhus commented:
I agree it should be documented: https://github.com/sindresorhus/is/commit/69bbf2d4bed61c0ceea3b38f93fb01fda94f52ee
